### PR TITLE
Add a new setting to add custom C# defines in the C# project

### PR DESCRIPTION
**Pull request by floriansimon1**
This is a pretty simple change that is fairly barebones but could become better over time. I would like to know if there's some interest for this from Godot developers?

### CODE REVIEW

**Comment by AThousandShips** [comment]
This is pretty much covered in a more extensive PR:
* https://github.com/godotengine/godot/pull/93539

I'd say this is a duplicate

**Comment by floriansimon1** [comment]
I could be wrong, but I don't think I can't use that other proposal to achieve what I'm going for.

I need the defines in the `.csproj` file for my text editor to know which compilation defines I'm using. I use this to bake in some debug tools and more generally enable some code paths in my code that I don't want in my final executable.

That's the part where I could be wrong, but I don't think my text editor will use the additional build arguments in the other PR when I'm just editing text. In my text editor, the code path that will be enabled is the one without any custom define.

**Comment by AThousandShips** [comment]
Then I'd suggest discussing this on the proposal to see the need for this and the overlap!

**Comment by floriansimon1** [comment]
I'll ask the author, but I don't expect there to be much overlap. We're for what seems to me is a completely different thing.

Thanks!

**Comment by AThousandShips** [comment]
I don't know enough C# to judge so I'll leave that to the dotnet team, but generally if a feature covers a particular demand with reasonable workarounds or adjustment it doesn't justify a separate feature or implementation, so unless it's completely unrelated it'd have to be discussed and evaluated for overlap and possible adjustments

**Comment by floriansimon1** [comment]
That makes sense!

**Comment by AThousandShips** [comment]
For the text editor side please open a proposal to discuss that side of things specifically, as it wasn't clear from the PR that the aim was for editor integration, more than the compilation side

File: modules/mono/editor/GodotTools/GodotTools.ProjectEditor/ProjectUtils.cs
Comment (floriansimon1) [inline modules/mono/editor/GodotTools/GodotTools.ProjectEditor/ProjectUtils.cs:1]: I should not pass `string.Empty` here!

File: modules/mono/editor/GodotTools/GodotTools/Internals/GodotSharpDirs.cs
Comment (floriansimon1) [inline modules/mono/editor/GodotTools/GodotTools/Internals/GodotSharpDirs.cs:1]: The additional static string is a leftoger that is no longer necessary.

**Comment by floriansimon1** [comment]
@AThousandShips thanks for your support today! I created the proposal and will wait for answers. I will still fix the PR, even if the PR gets rejected because I have the feeling some people might still want the patch. But I'll do so after work!

**Comment by paulloz** [comment]
Hello, and thank you for your contribution!

I am not sure that this is something we actually want in Godot. Contrary to other engines (at least to Unity), Godot do not take ownership of the `.csproj` file after its creation. We never overwrite, nor modify, manual changes made to the `.csproj` file by the user. Even when we regenerate it (e.g. when upgrading to a newer version of Godot). Thus, for me, this is about exposing a limited scope of functionality users already have access to (either through editing the XML itself, or through any C# IDE).
It makes me question why we'd expose an editor for constants, but not for everything else available in `.csproj`. And if we go for everything else, it makes me question why we're adding complexity (and maintainability cost) for an editor that most, if not all, C# users already have access to.

Side note, I think we could do a better job at explaining this in the documentation. Explicitly stating user changes to the `.csproj` files are possible, that they won't be overwritten, and maybe explain the default content of the file, and what shouldn't be manually removed.

**Comment by floriansimon1** [comment]
I can respect that point of view. If we add this, we might have to expose other options - things like separate assemblies, Nuget package references... The list goes on, and that's a lot of work.

I'm fine with having full ownership of the `.csproj`, but maybe Godot should not touch it without my permission. For instance, I have my own custom build of Godot, and something kept replacing the version of the Godot .NET SDK in my project to something that does not even exist without telling me anything every time I rebuilt my project. I finally managed to fix the issue by setting up a local Nuget repo with my custom build of the Godot SDK.

If the decision is to let devs manage their project files, I think they should know that they should. It's not immediately obvious, and editing the raw C# comes with its own quirks that I think we would do great to document. I can change my PR to do that if needed.

A victim of that sensible choice is user-friendliness. I had to spend multiple hours searching how to achieve what I wanted to do, and managed to break my game (the game got compiled without some Unity defines, and some references in my scenes broke and I even lost an animation tree configuration in my scene files!). Despite that, we get a lot of flexibility in exchange so I think it's worth it, even if it raises the barrier to entry slightly.

**Comment by paulloz** [comment]
> Godot should not touch it without my permission. For instance, I have my own custom build of Godot, and something kept replacing the version of the Godot .NET SDK in my project to something that does not even exist without telling me anything every time I rebuilt my project.

This should be the only thing Godot will ever modify in the `.csproj` file AFAIK.

> I finally managed to fix the issue by setting up a local Nuget repo with my custom build of the Godot SDK.

That is exactly what one should do when running a custom build.

> I can change my PR to do that if needed.

If so, that would be done on [godotengine/godot-docs](https://github.com/godotengine/godot-docs). I can open an issue over there to discuss what would make sense to document.

**Comment by floriansimon1** [comment]
@paulloz I would be glad to help. I can contribute the documentation in English and French if necessary! :smile:

**Pull request closed without merging**